src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,682 @@
+
+burnFrozenToken
+  :: forall store pm s. StorageC store pm
+  => (Address : Natural : store : s)
+  :-> store : s
+burnFrozenToken = do
+  dip $ do push frozenTokenId; pair
+  dig @2
+  debitFrom
+
+-- | Burn the 'slash_amount' calculated by 'cRejectedProposalReturnValue'
+-- If the 'slash_amount' is equal or bigger than the 'frozen' amount for that
+-- particular proposal, we only burn value equal to the 'frozen' amount.
+-- This is to avoid 'slash_amount' being abitrary large value, and we could burn
+-- the whole proposer frozen balance.
+burnSlashAmount
+  :: forall store pm s. StorageC store pm
+  => ("slash_amount" :! Natural) : Natural : Address : store : s
+  :-> ("slash_amount" :! Natural) : Natural : Address : store : s
+burnSlashAmount = do
+  duupX @2; duupX @2
+  swap; dip (fromNamed #slash_amount)
+  sub; isNat
+  ifSome
+    (do
+      drop
+      stackType @("slash_amount" :! Natural : Natural : Address : store : s)
+      dig @3; duupX @2; fromNamed #slash_amount -- Use the slash value
+      duupX @5
+      stackType @(Address : Natural : store : "slash_amount" :! Natural : Natural : Address : s)
+    )
+    (do
+      stackType @("slash_amount" :! Natural : Natural : Address : store : s)
+      dig @3; duupX @3 -- Use the frozen balance value
+      duupX @5
+      stackType @(Address : Natural : store : "slash_amount" :! Natural : Natural : Address : s)
+    )
+  burnFrozenToken
+  stackType @(store : "slash_amount" :! Natural : Natural : Address : s)
+  swap; dip $ do swap; dip swap
+
+unfreezeProposerToken
+  :: forall store pm. (NiceParameter pm, StorageC store pm)
+  => Config pm -> Bool -> [Proposal pm, store, ProposalKey, [Operation]]
+  :-> [Proposal pm, store, ProposalKey, [Operation]]
+unfreezeProposerToken Config{..} isAccepted = do
+
+  handleSlashed
+  stackType @[Natural, Address, store, Proposal pm, ProposalKey, [Operation]]
+
+  checkBalanceLessThanFrozenValue
+  unfreeze
+
+  stackType @[store, Proposal pm, ProposalKey, [Operation]]
+  swap
+  where
+    handleSlashed =
+      case isAccepted of
+        True -> do
+          stackType @[Proposal pm, store, ProposalKey, [Operation]]
+          getField #pProposerFrozenToken
+          dip $ getField #pProposer
+          dip $ dip swap
+          stackType @[Natural, Address, store, Proposal pm, ProposalKey, [Operation]]
+
+        False -> do
+          dup
+          cRejectedProposalReturnValue
+          stackType @["slash_amount" :! Natural, Proposal pm, store, ProposalKey, [Operation]]
+          dip $ do
+            getField #pProposerFrozenToken
+            dip $ getField #pProposer
+            dip $ dip swap
+          stackType @["slash_amount" :! Natural, Natural, Address, store, Proposal pm, ProposalKey, [Operation]]
+          burnSlashAmount
+
+          -- Calculate unfreeze amount
+          swap; dip (fromNamed #slash_amount)
+          sub; isNat
+          ifSome
+            (do
+              nop
```

Comment:
I think `nop` is redundant and then you don't need `do`.

Reply:
I've refactored this.